src/cc65/loadexpr.c

```diff
@@ -255,20 +263,82 @@ void LoadExpr (unsigned Flags, struct ExprDesc* Expr)
                 g_asr (F, Expr->BitOffs);
 
                 /* Since we have now shifted down, we could do char ops when the width fits in
-                ** a char, but we also need to clear the high byte since we've been using
-                ** CF_FORCECHAR up to now.
+                ** a char, but we also need to clear (or set) the high byte since we've been
+                ** using CF_FORCECHAR up to now.
                 */
+                unsigned Mask = (1U << Expr->BitWidth) - 1;
 
-                /* And by the width if the field doesn't end on a char or int boundary.
-                ** If it does end on a boundary, then zeros have already been shifted in,
-                ** but we need to clear the high byte for char.  g_and emits no code if the mask
-                ** is all ones.
+                /* To zero-extend, we will and by the width if the field doesn't end on a char or
+                ** int boundary.  If it does end on a boundary, then zeros will have already been
+                ** shifted in, but we need to clear the high byte for char.  g_and emits no code
+                ** if the mask is all ones.  This is here so the signed and unsigned branches can
+                ** use it.
                 */
+                unsigned ZeroExtendMask = 0;  /* Zero if we don't need to zero-extend. */
                 if (EndBit == CHAR_BITS) {
                     /* We need to clear the high byte, since CF_FORCECHAR was set. */
-                    g_and (BitFieldFullWidthFlags | CF_CONST, 0xFF);
+                    ZeroExtendMask = 0xFF;
                 } else if (EndBit != INT_BITS) {
-                    g_and (BitFieldFullWidthFlags | CF_CONST, (0x0001U << Expr->BitWidth) - 1U);
+                    ZeroExtendMask = (1U << Expr->BitWidth) - 1;
+                }
+
+                /* Handle signed bit-fields. */
+                if (IsSignSigned (Expr->Type)) {
+                    /* Push A, since the sign bit test will destroy it. */
+                    AddCodeLine ("pha");
```

Comment:
I see that `AddCodeLine` is mostly used in `codegen.c`. Let me know if these should be moved to something like `g_extendbitfield` in that file.

Reply:
Perhaps. The case for bitfields looks "inlined" a bit too much compared to others.

Reply:
@jmr: I presume you want to add another commit with the proposed modification.

Reply:
Sure. Let's do #1197 first, then I'll rebase on top of that.

Reply:
Done.